src/distances.jl

```diff
@@ -24,64 +84,50 @@ julia> Ellipsoidal([1.0,0.5], [π/2])
 julia> Ellipsoidal([1.0,0.5,0.5], [π/2,0.0,0.0])
 ```
 """
+
 struct Ellipsoidal{N,T} <: Metric
   dist::Mahalanobis{T}
 
-  function Ellipsoidal{N,T}(semiaxes, angles) where {N,T}
+  function Ellipsoidal{N,T}(semiaxes, angles; convention=:TaitBryanExtr) where {N,T}
     @assert length(semiaxes) == N "number of semiaxes must match spatial dimension"
     @assert all(semiaxes .> zero(T)) "semiaxes must be positive"
     @assert N ∈ [2,3] "dimension must be either 2 or 3"
 
-    # scaling matrix
-    Λ = Diagonal(one(T)./semiaxes.^2)
-
-    # rotation matrix
-    if N == 2
-      θ = angles[1]
-
-      cosθ = cos(θ)
-      sinθ = sin(θ)
+    rule = rules[convention]
 
-      P = [cosθ -sinθ
-           sinθ  cosθ]
+    # invert x and y if necessary
+    if rule.main == :y
+       semiaxes[1], semiaxes[2] = semiaxes[2], semiaxes[1]
     end
-    if N == 3
-      θxy, θyz, θzx = angles
-
-      cosxy = cos(θxy)
-      sinxy = sin(θxy)
-      cosyz = cos(θyz)
-      sinyz = sin(θyz)
-      coszx = cos(θzx)
-      sinzx = sin(θzx)
-
-      _1 = one(T)
-      _0 = zero(T)
-
-      Rxy = [cosxy -sinxy _0
-             sinxy  cosxy _0
-             _0     _0 _1]
 
-      Ryz = [_1    _0     _0
-             _0 cosyz -sinyz
-             _0 sinyz  cosyz]
+    # scaling matrix
+    Λ = Diagonal(one(T)./semiaxes.^2)
 
-      Rzx = [ coszx _0 sinzx
-             _0 _1    _0
-             -sinzx _0 coszx]
+    # convert to radian if necessary
+    angles = rule.radian ? angles : deg2rad.(angles)
 
-      P = Rzx*Ryz*Rxy
+    # rotation matrix
+    if N == 2
+      invert = (rule.motion[1]==:CW && !rule.extrinsic) || (rule.motion[1]==:CCW && rule.extrinsic)
+      θ = invert ? -angles[1] : angles[1]
+      P = angle_to_dcm(θ, 0.0, 0.0, rule.order)[1:2,1:2]
+    else # N == 3
+      for (i,sign) in enumerate(rule.motion)
+        invert = (sign==:CW && !rule.extrinsic) || (sign==:CCW && rule.extrinsic)
+        angles[i] = invert ? -angles[i] : angles[i]
+      end
+      P = angle_to_dcm(angles[1], angles[2], angles[3], rule.order)
     end
 
     # ellipsoid matrix
-    Q = P*Λ*P'
+    Q = rule.extrinsic ? P*Λ*P' : P'*Λ*P
```

Comment:
Just to confirm, this is still needed even after the sign flip above? Thanks for the careful PR, these angles are super tricky to get right.

Reply:
Also, it would be nice to double check the the resulting Q matrix is static after the multiplication with the diagonal.

Reply:
This sign flip is still necessary indeed
Checked about static matrix. Both P and Λ were as arrays (the first after reindexing it [1:N,1:N])
But after "force" StaticArrays, I got that:

```
ERROR: LoadError: MethodError: no method matching Distances.Mahalanobis(::StaticArrays.SArray{Tuple{2,2},Float64,2,4})
Closest candidates are:
  Distances.Mahalanobis(::Array{T,2}) where T
```

Seems to be some restriction in the `Distances.jl` (??)

Reply:
It seems that Mahalanobis and SqMahalanobis could be improved: https://github.com/JuliaStats/Distances.jl/blob/d547d3a78db98411a902fe5382243ece00bfdcc3/src/mahalanobis.jl#L3-L9

Do you mind submitting a PR to Distances.jl to add a parameter type like:

```julia
struct Mahalanobis{T,M<:AbstractMatrix{T}} <: Metric
    qmat::M
end
```
or even the cleaner:

```julia
struct Mahalanobis{M<:AbstractMatrix} <: Metric
    qmat::M
end
```

If they accept the improvement, we can work with static matrices, which should be much master. Otherwise, we will need to work a custom Mahalanobis in our side.

Reply:
Yeah
Well, I confess that these Julia parametric types/methods are still quite confusing for me.
But ok, I'll copy and paste your code in a issue, let's see how it goes